tests/training/optimizer_test.py

```diff
@@ -77,6 +80,171 @@ def test_optimizer_parameter_groups(self):
+                assert (
+                    name == "encoder"
+                ), "Optimizers were not initialized in the same order as parameter groups."
+                param_groups = inner_optimizer.param_groups
+                # The optimizer can have sub-groups with different options.
+                assert len(param_groups) == 2
+                for i, param_group in enumerate(param_groups):
+                    if i == 0:
+                        assert param_group["lr"] == 0.001
+                    elif i == 1:
+                        assert param_group["lr"] == 0.002
+            elif i == 2:
+                assert (
+                    name == "default"
+                ), "Optimizers were not initialized in the same order as parameter groups."
+                param_groups = inner_optimizer.param_groups
+                # Default group gets assigned any group without an 'optimizer_name' parameter in `parameter_groups`
+                # as well as any parameters which didn't match one of the given regexes.
+                assert len(param_groups) == 2
+
+    def test_optimizer_params(self):
+        optimizer_params = Params(
+            {
+                "type": "regex",
+                "optimizers": {
+                    "default": {"type": "adam", "lr": 1},
+                    "embedder": {"type": "adam", "lr": 2},
+                    "encoder": {"type": "adam", "lr": 3},
+                },
+                "parameter_groups": [
+                    [
+                        ["^text_field_embedder"],
+                        {
+                            "optimizer_name": "embedder",
+                            "betas": (0.9, 0.98),
+                            "lr": 2,
+                            "weight_decay": 0.01,
+                        },
+                    ],
+                    [["^encoder.*bias"], {"optimizer_name": "encoder", "lr": 0.001}],
+                    [["^encoder.*weight"], {"optimizer_name": "encoder", "lr": 0.002}],
+                    [["^tag_projection_layer.*.weight$"], {"lr": 5}],
+                ],
+            }
+        )
+
+        parameters = [[n, p] for n, p in self.model.named_parameters() if p.requires_grad]
+        optimizer = Optimizer.from_params(model_parameters=parameters, params=optimizer_params)
+
+        # When the RegexOptimizer is initialized, `optimizer.param_groups` stores the parameter groups.
+        # These parameter groups are assigned to their own optimizer by the RegexOptimizer.
+        # Check that changes to the parameters in optimizer.param_groups affect the parameters in
+        # optimizer._grouped_optimizers.
+        regex_optimizer_params = set()
+        regex_optimizer_grouped_optimizer_params = set()
+
+        for param_group in optimizer.param_groups:
+            # Each param_group should have optimizer options visible so they can be used by schedulers.
+            lr = param_group["lr"]
+            assert lr > 0
```

Comment:
This test still fails. On my machine, it's the only one that fails. The problem is that parameters that are not caught by any regex go into a default `param_group` with no parameters. This makes sense, because in that case the base parameters from the optimizer are used. It would also make sense if we didn't have that extra default `param_group` at all, but it looks like we've always had it, so I kept it.

My question is, @jbrry, why did this work in your code, and why do you think it should work? I don't think we guarantee that every `param_group` has an `lr` parameter, `RegexOptimizer` or not. That's not how `make_parameter_groups()` handles things. So I'm inclined to remove this test. But maybe you had it here for a reason?

Reply:
This test was related to the point about `learning_rate_schedulers` above, in that it needed to access the `"lr"` key from `self.optimizer.param_groups`.

It only worked in my code after [manually updating the `default` group's kwargs here](https://github.com/jbrry/allennlp/blob/fe463590d088f9c59654110bf09b65939caf75b0/allennlp/training/optimizers.py#L258-L263). I thought it should work so that the value would be visible to schedulers.

You're right, `make_parameter_groups()` only returns specified options in the dictionary (or empty for the `default` group) but after the Optimizer is initialized, the options make their way into the `self.optimizer.param_groups` so this must occur in the PyTorch source code (which is also relevant to the point above). So it might be worth trying to keep all relevant fields available to the outer optimizer, e.g. `self.optimizer.param_groups` so they can be accessed directly from there.

```
# Optimizer options for groups 0 and 1
group 0: {'lr': 0.002}
group 1: {}

# extra fields are then added to self.optimizer.param_groups after initialization
group 0: dict_keys(['params', 'lr', 'betas', 'eps', 'weight_decay', 'amsgrad', 'initial_lr'])
group 1: dict_keys(['params', 'lr', 'betas', 'eps', 'weight_decay', 'amsgrad', 'initial_lr'])

AdamOptimizer (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.9)
    eps: 1e-08
    initial_lr: 0.002
    lr: 0.0
    weight_decay: 0.0

Parameter Group 1
    amsgrad: False
    betas: (0.9, 0.9)
    eps: 1e-08
    initial_lr: 0.00123
    lr: 0.0
    weight_decay: 0.0
)
```

Reply:
This test is now passing since https://github.com/allenai/allennlp/commit/63ec4f3817491d748de58b0a2a42116bfd8d2b61.